python/mujinplanningclient/zmqsubscriber.py

```diff
@@ -0,0 +1,227 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012-2023 MUJIN Inc
+
+import threading
+
+from . import _
+from . import zmq
+from . import GetMonotonicTime
+from . import TimeoutError, UserInterrupt
+
+import logging
+log = logging.getLogger(__name__)
+
+
+class ZmqSubscriber(object):
+    """Subscriber that can handle ongoing subscriptions and automatic socket recreation.
+    """
+
+    _ctx = None # zmq context
+    _ctxown = None # created zmq context
+
+    _endpoint = None # zmq subscription endpoint
+    _endpointFn = None # function that returns the endpoint string to subscribe to, must be thread-safe
+    _callbackFn = None # function to call back when new message is received on the subscription socket
+
+    _socket = None # zmq socket
+    _lastReceivedTimestamp = 0 # when message was last received on this subscription
+    _timeout = 4.0 # beyond this number of seconds, the socket is considered dead and should be recreated
+    _checkpreemptfn = None # function for checking for preemptions
+
+    def __init__(self, endpoint, callbackFn=None, timeout=4.0, ctx=None, checkpreemptfn=None):
```

Comment:
let's have two fields: endpoint and endpointFn

Reply:
Done.